### PR TITLE
Initial code for the project

### PR DESCRIPTION
**Pull request by hasanheroglu**
This is the initial code for test-things. In development, I have worked with @egekorkan. 

The development has started at https://github.com/thingweb/device-sandbox but that repository will be closed/archived. Already existing devices could be hosted at plugfest.thinweb.io and we want to extend the repository with more devices in the future.

For feedback on the overall structure, please provide a comment and not a GitHub review (it can be the final comment of a GitHub review).

It is a big initial contribution but we wanted to show the overall structure. The actual code is small, there is a wrapper and test code.

### CODE REVIEW

**Comment by egekorkan** [comment]
Just for further clarification:
- My goal is to move every Thing we host to a dockerized setup
- The current Calculator example can be of course changed. A follow up PR will contribute the node-wot version of those Things.
- The utmost goal is to have multiple protocols that respect the protocol bindings where anyone can build a Consumer to test against these Things

File: LICENSE.md [hunk 1]
```diff
@@ -0,0 +1,277 @@
+Eclipse Public License - v 2.0
```
Comment (danielpeintner) [inline LICENSE.md:1]: General comment. The [LICENSE.md](https://github.com/eclipse-thingweb/node-wot/blob/master/LICENSE.md) file for node-wot looks different.

I don't have a strong opinion either way. I just think they should be aligned.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,66 @@
...
+The protocols you can currently test are:
+
+- HTTP
+- CoAP
+- MQTT
+- Modbus
+
+## Dependencies
+
+The project has several dependencies. Currently `JavaScript` and `Python` is used for simulating different devices. Every device has its own dependencies and they should be handled elegantly. For that reason, `Node.js` is used for JS code and `poetry` is used for Python code to run the scripts and handle the dependencies.
```
Comment (danielpeintner) [inline README.md:14]: I am not sure what you want to say with "elegantly"

File: README.md [hunk 3]
```diff
@@ -1,2 +1,66 @@
...
+For testing JavaScript Testing Framework `mocha` is used. Therefore, the tests are written in JavaScript.
+Every Thing should have its Thing Model and Thing Description validation test. 
+Thing Model validation test should be put under Thing's main directory and named as `tm.test.json`.
+Thing Description validation test should be put under protocol and programming language/framework's test directory and named as `td.test.json`.
+For Thing Description validation test, the device should boot up and to understand the device booted up without any error, a message `"ThingIsReady"` is expected to be prompted to the console by the device.
+
+## Adding a new Thing
+
+If you are going to add a different protocol for an existing Thing:
+1. Create a directory such as `<your_protocol>/<your_programming_language/your_framework>/`.
```
Comment (danielpeintner) [inline README.md:27]: Can it be that you missed to say that this directory should be under the "thing" directory?

File: README.md [hunk 4]
```diff
@@ -1,2 +1,66 @@
...
+2. Create your project files and write your code inside this directory.
+3. If you are adding a new programming language, please use a tidy dependency management tool for the programming language. Otherwise use already used tools and frameworks not to overcomplicate the project.
+4. Create your `test/` directory under your Thing's directory and add your test files there.  
+
+If you are going to add a different programming language/framework for an existing protocol:
+1. Create a directory such as `<existing_protocol>/<your_programming_language/your_framework>/`.
```
Comment (danielpeintner) [inline README.md:33]: same as before.. I think it should be under the thing directory

**Review by danielpeintner — APPROVED** [review]
I think it looks good. I added some comments (see below).

The main question I have is "why" we did not use node-wot for the exposed things. I see benefits of being neutral but somehow we partly implement it twice...

File: LICENSE.md (hunk 1 above)
Comment (egekorkan) [reply LICENSE.md:1]: Yes that is correct, we should simple copy it over

File: package.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "test-things",
+  "version": "1.0.0",
```
[the remark this answers is not in the document]
Comment (egekorkan) [reply package.json:3]: There is no intention to actually publish it so I think we can even remove it

**Comment by egekorkan** [comment]
> I think it looks good. I added some comments (see below).
> 
> The main question I have is "why" we did not use node-wot for the exposed things. I see benefits of being neutral but somehow we partly implement it twice...

So the node-wot Things will also come but my idea is to not couple the protocol implementation of a Thing to the changes in node-wot. There are many changes happening to node-wot that do not need to be applied to these Things. At the same time, a longer goal is to have these as stable as possible so that node-wot consumer side can be tested against them.

**Comment by danielpeintner** [comment]
> So the node-wot Things will also come but my idea is to not couple the protocol implementation of a Thing to the changes in node-wot. There are many changes happening to node-wot that do not need to be applied to these Things. At the same time, a longer goal is to have these as stable as possible so that node-wot consumer side can be tested against them.

Following you arguments I would say node-wot things should than *never* come. Having both does not make a lot of sense.. I think

File: LICENSE.md (hunk 1 above)
Comment (hasanheroglu) [reply LICENSE.md]: Fixed at [156d7be](https://github.com/eclipse-thingweb/test-things/pull/1/commits/156d7be2cc5b36e9d31e612a4f83c4abedef38fc)

File: README.md [hunk 6]
```diff
@@ -1,2 +1,66 @@
-# test-things
-Collection of Things that can be used for testing different IoT protocols, security mechanisms and interaction styles
+# Device Sandbox
```
[the remark this answers is not in the document]
Comment (hasanheroglu) [reply README.md]: Fixed at [156d7be](https://github.com/eclipse-thingweb/test-things/pull/1/commits/156d7be2cc5b36e9d31e612a4f83c4abedef38fc)

File: README.md (hunk 2 above)
Comment (hasanheroglu) [reply README.md]: Fixed at [156d7be](https://github.com/eclipse-thingweb/test-things/pull/1/commits/156d7be2cc5b36e9d31e612a4f83c4abedef38fc)

File: README.md (hunk 3 above)
Comment (hasanheroglu) [reply README.md]: Fixed at [156d7be](https://github.com/eclipse-thingweb/test-things/pull/1/commits/156d7be2cc5b36e9d31e612a4f83c4abedef38fc)

File: README.md (hunk 4 above)
Comment (hasanheroglu) [reply README.md]: Fixed at [156d7be](https://github.com/eclipse-thingweb/test-things/pull/1/commits/156d7be2cc5b36e9d31e612a4f83c4abedef38fc)

File: package.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (hasanheroglu) [reply package.json]: Removed at [156d7be](https://github.com/eclipse-thingweb/test-things/pull/1/commits/156d7be2cc5b36e9d31e612a4f83c4abedef38fc)

File: scripts/installDependencies.sh [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+    # Install dependencies for poetry 
+    if [ -f "poetry.lock" ]; then
+        poetry install
+    fi
+
+    cd $current_path
+done
+
+exit 0
```
[the remark this answers is not in the document]
Comment (hasanheroglu) [reply scripts/installDependencies.sh]: Fixed at [156d7be](https://github.com/eclipse-thingweb/test-things/pull/1/commits/156d7be2cc5b36e9d31e612a4f83c4abedef38fc)

File: things/calculator/coap/js/main.js [hunk 8]
```diff
@@ -0,0 +1,138 @@
+const { parseArgs } = require("node:util");
```
[the remark this answers is not in the document]
Comment (hasanheroglu) [reply things/calculator/coap/js/main.js]: I think after @egekorkan's explanation below about using node-wot, this is not needed.

**Comment by egekorkan** [comment]
Talked with @danielpeintner as well and we can merge it. I will create a discussion for the overall management of such Things

**Pull request merged**